### PR TITLE
Patch: Invalid Datetime, toArray errors + Mapping: Discharge Summary missing

### PR DESCRIPTION
**Pull request by jonahkaye**
Refs: #[1449](https://github.com/metriport/metriport-internal/issues/1449)

### Description

- [x] patch invalid datetimes to just log if invalid.
- [x] patch toArray errors
- [x] added Discharge Summary parsing
- [x] Edits to e2e testing local to count resources without inserting into fhir

### Testing

- Local
  - [x] local fhir mr gen
  - [x] test on a dozen patients 

### Release Plan

- [x] this points to master
- [ ] Merge this

### CODE REVIEW

**Comment by jonahkaye** [comment]
This file: https://s3.console.aws.amazon.com/s3/object/metriport-medical-documents?region=us-west-1&bucketType=general&prefix=a9ac975c-2da8-4c56-ad67-4a3fef79252e/018d12dd-d007-7ead-a29d-e1ee6aabc6f4/a9ac975c-2da8-4c56-ad67-4a3fef79252e_018d12dd-d007-7ead-a29d-e1ee6aabc6f4_018d13f1-8854-756c-9220-fd758bcbb523.xml

caused this error: `Response body: {"error":{"code":"BadRequest","message":"Unable to parse input data. Error: Non-whitespace before first tag.\nLine: 0\nColumn: 1\nChar: d"} ` but the file itself isnt an xml

File: packages/fhir-converter/src/templates/cda/Resources/FamilyMemberHistory.hbs [hunk 1]
```diff
@@ -15,7 +15,7 @@
         "bornString":"{{familyEntry.subject.relatedSubject.subject.birthTime.value}}",
         "condition":
         [
-            {{#each familyEntry.component}}
+            {{#each (toArray familyEntry.component)}}
```
Comment (jonahkaye) [inline packages/fhir-converter/src/templates/cda/Resources/FamilyMemberHistory.hbs:18]: resulted in us losing the actual condition in the family history!

File: packages/fhir-converter/src/lib/handlebars-converter/handlebars-helpers.js [hunk 2]
```diff
@@ -849,7 +849,7 @@ module.exports.external = [
       try {
         return getDateTime(dateTimeString);
       } catch (err) {
-        throw `helper "formatAsDateTime" : ${err}`;
+        console.log(`helper "formatAsDateTime" : ${err}`);
```
Comment (jonahkaye) [inline packages/fhir-converter/src/lib/handlebars-converter/handlebars-helpers.js:852]: no need to throw error. question is why the regex replacements in preprocessing step are not catching these cases 🤔

**Review by jonahkaye — COMMENTED** [review]
```
Found 115 XML files.
Converting 115 files, 10 at a time...
Converted 115 files in 1772 ms.
Removing all partitions from the FHIR Server...
Creating tenant @ FHIRServer w/ {"cxId":"018d5b76-e517-730a-8f25-7d48ea8a47c5","organizationNumber":524834}
Found 115 JSON files.
Inserting 115 files, 1 at a time...
Inserted 115 files in 20412 ms.
Resources:  {
  AllergyIntolerance: 2,
  Communication: 12,
  Composition: 115,
  Condition: 134,
  Coverage: 3,
  Device: 45,
  DiagnosticReport: 165,
  Encounter: 236,
  FamilyMemberHistory: 14,
  Immunization: 12,
  MedicationAdministration: 12,
  MedicationRequest: 5,
  MedicationStatement: 5,
  Observation: 626,
  Procedure: 34,
  RelatedPerson: 1
}
Total resources: 1421
Total time: 20478 ms / 0.3413 min
jonahkaye@Jonahs-MBP utils % ts-node src/fhir-converter/e2e-test.ts --use-fhir-server --cleanup
(node:52330) NOTE: We are formalizing our plans to enter AWS SDK for JavaScript (v2) into maintenance mode in 2023.

Please migrate your code to use AWS SDK for JavaScript (v3).
For more information, check the migration guide at https://a.co/7PzMCcy
(Use `node --trace-warnings ...` to show where the warning was created)
Running End-to-End tests for the FHIR Converter (useFhirServer: true, cleanup: true) - started at 2024-01-30T17:45:25.151Z
Log folder: runs/fhir-converter-e2e/2024-01-30T17:45:25.048Z
Found 84 XML files.
Converting 84 files, 10 at a time...
Converted 84 files in 1608 ms.
Removing all partitions from the FHIR Server...
Creating tenant @ FHIRServer w/ {"cxId":"018d5b79-cb37-775e-ae57-f9cb40488814","organizationNumber":575803}
Found 84 JSON files.
Inserting 84 files, 1 at a time...
Inserted 84 files in 20473 ms.
Resources:  {
  AllergyIntolerance: 3,
  Communication: 11,
  Composition: 84,
  Condition: 107,
  Coverage: 1,
  Device: 33,
  DiagnosticReport: 254,
  Encounter: 170,
  FamilyMemberHistory: 10,
  Immunization: 6,
  MedicationAdministration: 15,
  MedicationRequest: 6,
  MedicationStatement: 6,
  Observation: 625,
  Procedure: 70,
  RelatedPerson: 1
}
Total resources: 1402
```

File: packages/fhir-converter/src/templates/cda/Sections/Notes.hbs [hunk 3]
```diff
@@ -1,12 +1,34 @@
...
+                            {{>Resources/Practitioner.hbs practitioner=note.act.author.assignedAuthor ID=practitionerId.Id}},
+                            {{>References/DiagnosticReport/performer.hbs ID=(generateUUID (toJsonString note.act)) REF=(concat 'Practitioner/' practitionerId.Id)}},
+                        {{/with}}
+                        {{>Resources/Organization.hbs org=note.act.author.assignedAuthor.representedOrganization ID=(generateUUID (toJsonString note.act.author.assignedAuthor.representedOrganization))}},
+                        {{>References/DiagnosticReport/performer.hbs ID=(generateUUID (toJsonString note.act)) REF=(concat 'Organization/' (generateUUID (toJsonString note.act.author.assignedAuthor.representedOrganization)))}},
+                    {{/if}}   
+                {{/each}}
+            {{/each}}
+        {{/with}}
+        {{!-- for discharge summarries. we can onbviously refactor this or put it in its own section or just modify the getAllCdaSectionsByTemplateId to take multiple ids --}}
```
Comment (jonahkaye) [inline packages/fhir-converter/src/templates/cda/Sections/Notes.hbs:23]: added this. We werent picking up discharge summaries before and there were several CDAs I was looking at where they had full notes in them that we were missing

File: packages/fhir-converter/src/templates/cda/Sections/Notes.hbs [hunk 4]
```diff
@@ -1,12 +1,34 @@
...
+        {{#with (getAllCdaSectionsByTemplateId msg '1.3.6.1.4.1.19376.1.5.3.1.3.5') as |sections|}}
```
Comment (Goncharo) [inline packages/fhir-converter/src/templates/cda/Sections/Notes.hbs:24]: shouldn't we put this in its own "Hospital Course Section".hbs or something?

why would we not want to pull discharge summaries if the `2.16.840.1.113883.10.20.22.2.65` section doesn't exist?

**Review by Goncharo — CHANGES_REQUESTED** [review]

File: packages/fhir-converter/src/templates/cda/Sections/Notes.hbs (hunk 4 above)
Comment (jonahkaye) [reply packages/fhir-converter/src/templates/cda/Sections/Notes.hbs]: I refactored to address this issue

**Review by jonahkaye — COMMENTED** [review]
Resource Count before inserting into FHIR. Much larger because FHIR does some dedup:

```
Resources: {
  "AllergyIntolerance": 819,
  "Composition": 9445,
  "Condition": 17018,
  "Consent": 53,
  "Coverage": 744,
  "Device": 9207,
  "FamilyMemberHistory": 774,
  "Medication": 9892,
  "MedicationStatement": 5585,
  "Observation": 151097,
  "Organization": 20514,
  "Patient": 9445,
  "Practitioner": 40061,
  "RelatedPerson": 12240,
  "DiagnosticReport": 30906,
  "Encounter": 18582,
  "Location": 17626,
  "Procedure": 9603,
  "Immunization": 2456,
  "MedicationRequest": 4975,
  "MedicationAdministration": 5387,
  "Communication": 1525,
  "ServiceRequest": 1,
  "Goal": 547
}
Total: 378502
```

Delta to develop:

- [x] DiagnosticReport: 27459 => 30906
- [x] Organization: 20506 => 20514
- [x] Practitioner: 40036 => 40061

**Review by Goncharo — APPROVED** [review]
:shipit: - conditional approval based on running the insert test on this PR

File: packages/fhir-converter/src/templates/cda/Sections/Notes.hbs [hunk 5]
```diff
@@ -1,7 +1,8 @@
-{{#if (contains (toString (toJsonString msg)) '2.16.840.1.113883.10.20.22.2.65')}}
-    {{#with (getAllCdaSectionsByTemplateId msg '2.16.840.1.113883.10.20.22.2.65') as |sections|}}
+{{#if (contains (toString (toJsonString msg)) '2.16.840.1.113883.10.20.22.2.65' '1.3.6.1.4.1.19376.1.5.3.1.3.5')}}
+    {{#with (getAllCdaSectionsByTemplateId msg '2.16.840.1.113883.10.20.22.2.65' '1.3.6.1.4.1.19376.1.5.3.1.3.5') as |sections|}}
+            {{!-- sections here is already an array so this is not a bug --}}
             {{#each sections}}
-                {{#each (toArray 2_16_840_1_113883_10_20_22_2_65.entry) as |note|}}
+                {{#each (multipleToArray this.[2_16_840_1_113883_10_20_22_2_65].entry this.[1_3_6_1_4_1_19376_1_5_3_1_3_5].entry) as |note|}}
```
Comment (jonahkaye) [inline packages/fhir-converter/src/templates/cda/Sections/Notes.hbs:5]: DiagnosticReport: 27459 => 30906
Organization: 20506 => 20514
Practitioner: 40036 => 40061

More diagnostic reports, and each diagnostic reports entails the creation of more practitioners and organizations.

**Review by jonahkaye — COMMENTED** [review]
```
Resources:  {
  AllergyIntolerance: 743,
  Communication: 423,
  Composition: 9411,
  Condition: 13091,
  Consent: 44,
  Coverage: 558,
  Device: 3830,
  DiagnosticReport: 24970,
  Encounter: 18460,
  FamilyMemberHistory: 725,
  Goal: 45,
  Immunization: 2123,
  MedicationAdministration: 2687,
  MedicationRequest: 3428,
  MedicationStatement: 4691,
  Observation: 79392,
  Procedure: 8915,
  RelatedPerson: 269,
  ServiceRequest: 1,
  Practitioner: 8203,
  Organization: 1552,
  Medication: 2388,
  Location: 3675
}
Total resources w/o Practitioner, Org, Location: 176194
Total resources: 189624
```

**Pull request merged**